src/main/java/org/optaplanner/springboottaskassigning/TaskAssigningSolverManagerController.java

```diff
@@ -16,42 +16,68 @@
 
 package org.optaplanner.springboottaskassigning;
 
+import java.util.NoSuchElementException;
+
 import org.optaplanner.core.api.score.Score;
 import org.optaplanner.springboottaskassigning.domain.TaskAssigningSolution;
-import org.optaplanner.springboottaskassigning.solver.SolverManager;
 import org.optaplanner.springboottaskassigning.solver.SolverStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
 import org.springframework.web.bind.annotation.GetMapping;
 import org.springframework.web.bind.annotation.PathVariable;
 import org.springframework.web.bind.annotation.PostMapping;
 import org.springframework.web.bind.annotation.RequestBody;
 import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.server.ResponseStatusException;
 
 @RestController
-@RequestMapping("solvers")
+@RequestMapping("/tenants/{tenantId}/solver")
 public class TaskAssigningSolverManagerController {
 
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
     @Autowired
-    private SolverManager<TaskAssigningSolution> solverManager;
+    private TaskAssigningSolverManagerService solverManagerService;
 
-    @PostMapping("{tenantId}")
-    public void solve(@PathVariable Comparable<?> tenantId, @RequestBody TaskAssigningSolution planningProblem) {
-        solverManager.solve(tenantId, planningProblem);
+    @PostMapping
+    public void solve(@PathVariable Long tenantId, @RequestBody TaskAssigningSolution planningProblem) {
+        logger.debug("POST /tenants/{}/solver", tenantId);
+        solverManagerService.solve(tenantId, planningProblem);
     }
 
-    @GetMapping("{tenantId}/bestSolution")
-    public TaskAssigningSolution bestSolution(@PathVariable Comparable<?> tenantId) {
-        return solverManager.getBestSolution(tenantId);
+    @GetMapping("/bestSolution")
+    public TaskAssigningSolution bestSolution(@PathVariable Long tenantId) {
+        logger.debug("GET /tenants/{}/solver/bestSolution", tenantId);
+        try {
+            return solverManagerService.bestSolution(tenantId);
+        } catch (NoSuchElementException e) {
```

Comment:
@rsynek This a is a good question:
If nothing is find, does it:
- A) return null (like Map.get(key))
- B) throw NoSuchElementException
- C) block until there is such a thing

The first thing we need, is a list of JDK APi's that throw NoSuchElementException.

Reply:
Also "bestSolution(tenantId)" should probably be getBestSolution(tenantId)

Reply:
ad A) I would rather avoid returning nulls in public methods, it leaks outside the API.
ad B) might be good if you really consider it exceptional state, otherwise let's use Optional
ad C) if blocking, only with a timeout

Reply:
C) Unless we have a strong feedback from users, I would rather return immediately with a helpful message and give users the option of blocking and trying again or doing something else and trying again later.
A,B) NoSuchElementException is thrown from `solverManagerService.bestSolution(tenantId)` because it calls `Optional<Solution_>.get()` without checking `optaionl.isPresent()`, the alternative is to check `isPresent()` and return `NULL` otherwise, but at least in this case the developer is aware of the possibility that `solverManager.getBestSolution()` might return null.

Reply:
I took a look at the code in IDE and here is my view on this:
- `SolverManager` returns `Optional`, that is right.
- `TaskAssigningSolverManagerService` just delegates on the `SolverManager`. It's still a Java API and we cannot guarantee it will always return a non-null object. Thus, I would keep `Optional` here as a return type, too.
- `TaskAssigningSolverManagerController` should check the `Optional` instead of catching the `NoSuchElementException` and return either the value or a self-explanatory answer.

Why:
`Optional` should be passed to other classes until there is a need to transform it into a different data type. Only the terminal operation, that really needs to do something with the value, should check the presence of the value in `Optional` and react properly.
In someone will be dealing with the `TaskAssigningSolverManagerService` directly, the contract is clear an he won't have to check for NSEE.